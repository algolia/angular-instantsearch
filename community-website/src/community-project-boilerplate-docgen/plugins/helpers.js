const highlight = require('./syntaxHighlighting.js');
const md = require('./mdRenderer.js');

// this plugin provides ATM one helper to easily compute the publicPath of assets
module.exports = function helpers(filenames, metalsmith, cb) {
  // eslint-disable-next-line no-param-reassign
  metalsmith.metadata().h = {
    markdown(src) {
      return md.render(src);
    },
    highlight(src, opts) {
      const lang = opts && opts.lang;
      return highlight(src, lang);
    },
    maybeActive(navPath, singlePathOrArrayOfPaths) {
      const pathsToTest = [].concat(singlePathOrArrayOfPaths);
      return pathsToTest.some(pathToTest => navPath.indexOf(pathToTest) === 0)
        ? 'active'
        : '';
    },
  };

  cb();
};
