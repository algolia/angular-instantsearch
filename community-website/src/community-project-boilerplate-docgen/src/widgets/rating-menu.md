---
title: Widgets - Rating Menu
layout: widget.pug
---

# Rating Menu

## Description

RatingMenu is used for displaying grade like filters. The values are normalized within boundaries.

## Options

| Attributes           | Type       | Description
| -                    | -          | -
| `attribute`          | `string`   | Name of the attribute for faceting
| `andUpLabel?`        | `string`   | Label to display after the stars
| `max?`               | `number`   | The maximum rating value
| `autoHideContainer?` | `boolean`  | Hides the rating menu if there's no item to display

## Code example

```html
<ng-ais-rating-menu
  attribute="rating"
  [max]="5"
>
</ng-ais-rating-menu>
```
