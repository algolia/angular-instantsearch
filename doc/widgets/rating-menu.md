## `<ng-ais-rating-menu />`

RatingMenu is used for displaying grade like filters. The values are normalized within boundaries.

| Attributes           | Type       | Description
| -                    | -          | -
| `attributeName`      | `string`   | Name of the attribute for faceting
| `andUpLabel?`        | `string`   | Label to display after the stars
| `max?`               | `number`   | The maximum rating value
| `header?`            | `string`   | Displays text before widget
| `footer?`            | `string`   | Displays text after widget
| `autoHideContainer?` | `boolean`  | Hides the rating menu if there's no item to display
