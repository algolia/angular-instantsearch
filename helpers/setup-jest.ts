import "jest-preset-angular";

// stupid workaround for TS
import * as helper from "algoliasearch-helper";
require("algoliasearch-helper").default = helper;

const mock = () => {
  let storage = {};
  return {
    getItem(key) {
      return storage[key] || null;
    },
    setItem(key, value) {
      storage[key] = value;
    },
    removeItem(key) {
      delete storage[key];
    },
    clear() {
      storage = {};
    }
  };
};

Object.defineProperty(window, "localStorage", { value: mock() });
Object.defineProperty(window, "sessionStorage", { value: mock() });
Object.defineProperty(window, "getComputedStyle", {
  value: () => ["-webkit-appearance"]
});
