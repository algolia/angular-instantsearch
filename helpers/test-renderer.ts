import { Component, ViewChild } from "@angular/core";
import { TestBed } from "@angular/core/testing";

import { NgAisInstantSearchModule } from "../src/instantsearch/instantsearch.module";

// mock component, dont create a real instantearch instance
jest.mock("../src/base-widget");
jest.mock("../src/instantsearch/instantsearch");

export function createRenderer({
  template,
  TestedWidget,
  defaultState,
  additionalImports,
  additionalDeclarations
}: {
  template: string;
  TestedWidget: any;
  additionalImports?: any[];
  additionalDeclarations?: any[];
  defaultState?: {};
}) {
  return function(state?: {}, firstRender = false) {
    return render(
      {
        template,
        TestedWidget,
        additionalImports,
        additionalDeclarations,
        state: state ? { ...(defaultState || {}), ...state } : undefined
      },
      firstRender
    );
  };
}

function render(
  {
    template,
    TestedWidget,
    additionalImports,
    additionalDeclarations,
    state
  }: {
    template: string;
    TestedWidget: any;
    additionalImports?: any[];
    additionalDeclarations?: any[];
    state?: {};
  },
  firstRender = false
) {
  @Component({
    template: `
      <ng-ais-instantsearch>
        ${template}
      </ng-ais-instantsearch>
    `
  })
  class TestContainer {
    @ViewChild(TestedWidget) testedWidget;
  }

  TestBed.configureTestingModule({
    declarations: [
      TestContainer,
      TestedWidget,
      ...(additionalDeclarations || [])
    ],
    imports: [NgAisInstantSearchModule.forRoot(), ...(additionalImports || [])]
  });

  const fixture = TestBed.createComponent(TestContainer);
  fixture.detectChanges();

  if (state) {
    fixture.componentInstance.testedWidget.updateState(state, firstRender);
    fixture.detectChanges();
  }

  return fixture;
}
