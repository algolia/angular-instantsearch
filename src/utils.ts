export function bem(widgetName: string) {
  const cx = function(element?: string, subElement?: string) {
    if (element) {
      const scoppedWidgetName = `ais-${widgetName}-${element}`;

      // output `ais-Widget-Header|Body|Footer ais-Header|Body|Footer`
      if (element === "header" || element === "body" || element === "footer") {
        const nonScoppedWidgetName = `ais-${element}`;
        return `${scoppedWidgetName} ${nonScoppedWidgetName}`;
      }

      // output `ais-Widget-Xyz--abc`
      if (subElement) {
        return `${scoppedWidgetName}--${subElement}`;
      }

      // output `ais-Widget-Xyz`
      return scoppedWidgetName;
    }

    // output `ais-Widget`
    return `ais-${widgetName}`;
  };
  return cx;
}

export function parseNumberInput(input?: number | string) {
  return typeof input === "string" ? parseInt(input, 10) : input;
}
