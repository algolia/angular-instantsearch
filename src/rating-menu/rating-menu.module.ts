import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { NgAisRatingMenu } from './rating-menu';

@NgModule({
  declarations: [NgAisRatingMenu],
  entryComponents: [NgAisRatingMenu],
  exports: [NgAisRatingMenu],
  imports: [CommonModule],
})
export class NgAisRatingMenuModule {}
