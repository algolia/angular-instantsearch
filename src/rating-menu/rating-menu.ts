import { Component, Input, Inject, forwardRef } from '@angular/core';

import { connectRatingMenu } from 'instantsearch.js/es/connectors';
import { BaseWidget } from '../base-widget';
import { NgAisInstantSearch } from '../instantsearch/instantsearch';
import { noop, parseNumberInput } from '../utils';

export type RatingMenuItem = {
  count: number;
  isRefined: boolean;
  name: string;
  value: string;
  stars: boolean[];
};

export type RatingMenuState = {
  createURL: Function;
  hasNoResults: boolean;
  items: RatingMenuItem[];
  refine: Function;
};

@Component({
  selector: 'ais-rating-menu',
  template: `
    <div
      [class]="cx()"
      *ngIf="!isHidden"
    >
      <svg style="display:none;">
        <symbol
          id="ais-StarRating-starSymbol"
          viewBox="0 0 24 24"
          width="24"
          height="24"
        >
          <path d="M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z"/>
        </symbol>
        <symbol
          id="ais-StarRating-starEmptySymbol"
          viewBox="0 0 24 24"
          width="24"
          height="24"
        >
          <path d="M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z"/>
        </symbol>
      </svg>

      <ul [class]="cx('list')">
        <li
          *ngFor="let item of state.items"
          [class]="getItemClass(item)"
          (click)="handleClick($event, item.value)"
        >
          <a
            href="{{state.createURL(item.value)}}"
            [class]="cx('link')"
            (click)="handleClick($event, item.value)"
          >
            <svg
              *ngFor="let star of item.stars"
              [ngClass]="cx('starIcon')"
              aria-hidden="true"
            >
              <use
                *ngIf="star"
                xlink:href="#ais-StarRating-starSymbol"
              >
              </use>

              <use
                *ngIf="!star"
                xlink:href="#ais-StarRating-starEmptySymbol"
              >
              </use>
            </svg>

            <span [class]="cx('label')" aria-hidden="true">{{andUpLabel}}</span>
            <span [class]="cx('count')">{{item.count}}</span>
          </a>
        </li>
      </ul>
    </div>
  `,
})
export class NgAisRatingMenu extends BaseWidget {
  // rendering options
  @Input() public andUpLabel: string = '& Up';

  // instance options
  @Input() public attribute: string;
  @Input() public max?: number;

  public state: RatingMenuState = {
    createURL: noop,
    hasNoResults: false,
    items: [],
    refine: noop,
  };

  get isHidden() {
    return this.state.items.length === 0 && this.autoHideContainer;
  }

  constructor(
    @Inject(forwardRef(() => NgAisInstantSearch))
    public instantSearchParent: any
  ) {
    super('RatingMenu');
  }

  public ngOnInit() {
    this.createWidget(connectRatingMenu, {
      attribute: this.attribute,
      max: parseNumberInput(this.max),
    });
    super.ngOnInit();
  }

  public handleClick(event: MouseEvent, value: string) {
    event.preventDefault();
    event.stopPropagation();

    this.state.refine(value);
  }
}
