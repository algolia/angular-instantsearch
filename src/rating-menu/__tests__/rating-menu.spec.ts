import { createRenderer } from "../../../helpers/test-renderer";
import { NgAisRatingMenu } from "../rating-menu";

import { bem } from "../../utils";

const cx = bem("StarRating");

const defaultState = {
  createURL: jest.fn(),
  hasNoResults: false,
  items: [
    {
      name: "4",
      stars: [true, true, true, true, false],
      count: 100,
      isRefined: true
    },
    { name: "3", stars: [true, true, true, false, false], count: 100 },
    { name: "2", stars: [true, true, false, false, false], count: 100 },
    { name: "1", stars: [true, false, false, false, false], count: 100 }
  ],
  refine: jest.fn()
};

const render = createRenderer({
  defaultState,
  template: "<ais-rating-menu></ais-rating-menu>",
  TestedWidget: NgAisRatingMenu
});

describe("StarRating", () => {
  it("should render without state", () => {
    const fixture = render();
    expect(fixture).toMatchSnapshot();
  });

  it("should render with state", () => {
    const fixture = render({});
    expect(fixture).toMatchSnapshot();
  });

  it("should be hidden with autoHideContainer", () => {
    const fixture = render({ items: [] });
    fixture.componentInstance.testedWidget.autoHideContainer = true;
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });
});
