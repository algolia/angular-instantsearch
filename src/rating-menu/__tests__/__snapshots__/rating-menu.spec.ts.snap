// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarRating should be hidden with autoHideContainer 1`] = `
<ng-component
  testedWidget={[Function NgAisRatingMenu]}
>
  <ais-instantsearch>
    <ais-rating-menu>
      
    </ais-rating-menu>
  </ais-instantsearch>
</ng-component>
`;

exports[`StarRating should render with state 1`] = `
<ng-component
  testedWidget={[Function NgAisRatingMenu]}
>
  <ais-instantsearch>
    <ais-rating-menu>
      
      <div
        class="ais-RatingMenu"
      >
        <svg
          style="display:none;"
        >
          <symbol
            height="24"
            id="ais-StarRating-starSymbol"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z"
            />
          </symbol>
          <symbol
            height="24"
            id="ais-StarRating-starEmptySymbol"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z"
            />
          </symbol>
        </svg>
        <ul
          class="ais-RatingMenu-list"
        >
          
          <li
            class="ais-RatingMenu-item ais-RatingMenu-item--selected"
          >
            <a
              class="ais-RatingMenu-link"
              href=""
            >
              
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <span
                aria-hidden="true"
                class="ais-RatingMenu-label"
              >
                & Up
              </span>
              <span
                class="ais-RatingMenu-count"
              >
                100
              </span>
            </a>
          </li>
          <li
            class="ais-RatingMenu-item"
          >
            <a
              class="ais-RatingMenu-link"
              href=""
            >
              
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <span
                aria-hidden="true"
                class="ais-RatingMenu-label"
              >
                & Up
              </span>
              <span
                class="ais-RatingMenu-count"
              >
                100
              </span>
            </a>
          </li>
          <li
            class="ais-RatingMenu-item"
          >
            <a
              class="ais-RatingMenu-link"
              href=""
            >
              
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <span
                aria-hidden="true"
                class="ais-RatingMenu-label"
              >
                & Up
              </span>
              <span
                class="ais-RatingMenu-count"
              >
                100
              </span>
            </a>
          </li>
          <li
            class="ais-RatingMenu-item"
          >
            <a
              class="ais-RatingMenu-link"
              href=""
            >
              
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                <use
                  xlink:href="#ais-StarRating-starSymbol"
                />
                
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <svg
                aria-hidden="true"
                class="ais-RatingMenu-starIcon"
                ng-reflect-ng-class="ais-RatingMenu-starIcon"
              >
                
                
                <use
                  xlink:href="#ais-StarRating-starEmptySymbol"
                />
              </svg>
              <span
                aria-hidden="true"
                class="ais-RatingMenu-label"
              >
                & Up
              </span>
              <span
                class="ais-RatingMenu-count"
              >
                100
              </span>
            </a>
          </li>
        </ul>
      </div>
    </ais-rating-menu>
  </ais-instantsearch>
</ng-component>
`;

exports[`StarRating should render without state 1`] = `
<ng-component
  testedWidget={[Function NgAisRatingMenu]}
>
  <ais-instantsearch>
    <ais-rating-menu>
      
      <div
        class="ais-RatingMenu"
      >
        <svg
          style="display:none;"
        >
          <symbol
            height="24"
            id="ais-StarRating-starSymbol"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12 .288l2.833 8.718h9.167l-7.417 5.389 2.833 8.718-7.416-5.388-7.417 5.388 2.833-8.718-7.416-5.389h9.167z"
            />
          </symbol>
          <symbol
            height="24"
            id="ais-StarRating-starEmptySymbol"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12 6.76l1.379 4.246h4.465l-3.612 2.625 1.379 4.246-3.611-2.625-3.612 2.625 1.379-4.246-3.612-2.625h4.465l1.38-4.246zm0-6.472l-2.833 8.718h-9.167l7.416 5.389-2.833 8.718 7.417-5.388 7.416 5.388-2.833-8.718 7.417-5.389h-9.167l-2.833-8.718z"
            />
          </symbol>
        </svg>
        <ul
          class="ais-RatingMenu-list"
        >
          
        </ul>
      </div>
    </ais-rating-menu>
  </ais-instantsearch>
</ng-component>
`;
